table/tables/partition.go

```diff
@@ -180,29 +223,48 @@ func partitionRecordKey(pid int64, handle int64) kv.Key {
 func (t *partitionedTable) locatePartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int64, error) {
 	var err error
 	var isNull bool
-	partitionExprs := t.partitionExpr.UpperBounds
-	idx := sort.Search(len(partitionExprs), func(i int) bool {
+	idx := 0
+	switch t.meta.Partition.Type {
+	case model.PartitionTypeRange:
+		partitionExprs := t.partitionExpr.UpperBounds
+		idx = sort.Search(len(partitionExprs), func(i int) bool {
+			var ret int64
+			ret, isNull, err = partitionExprs[i].EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
+			if err != nil {
+				return true // Break the search.
+			}
+			if isNull {
+				// If the column value used to determine the partition is NULL, the row is inserted into the lowest partition.
+				// See https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-handling-nulls.html
+				return true // Break the search.
+			}
+			return ret > 0
+		})
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		if isNull {
+			idx = 0
+		}
+		if idx < 0 || idx >= len(partitionExprs) {
+			// The data does not belong to any of the partition?
+			return 0, errors.Trace(table.ErrTrgInvalidCreationCtx)
+		}
+
+	case model.PartitionTypeHash:
 		var ret int64
-		ret, isNull, err = partitionExprs[i].EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
+		ret, isNull, err = t.partitionExpr.Expr.EvalInt(ctx, chunk.MutRowFromDatums(r).ToRow())
 		if err != nil {
-			return true // Break the search.
+			return 0, errors.Trace(err)
 		}
 		if isNull {
-			// If the column value used to determine the partition is NULL, the row is inserted into the lowest partition.
-			// See https://dev.mysql.com/doc/mysql-partitioning-excerpt/5.7/en/partitioning-handling-nulls.html
-			return true // Break the search.
+			idx = 0
+		} else {
+			if ret < 0 {
```

Comment:
What's the behavior in MySQL when it's minus ? Is there any test cover that ?

Reply:
I have not found any information about this in mysql document. 
Here is a test is mysql. I think the current behavior in this PR is same with mysql.
```sql
mysql root@127.0.0.1:test> create table th (
                        ->   a int,
                        ->   b int
                        -> ) partition by hash(a) partitions 5;
Query OK, 0 rows affected
Time: 0.027s
mysql root@127.0.0.1:test> insert into th values (-1,-1),(-2,-2),(-3,-3),(-4,-4),(-5,-5),(-6,-6),(-7,-7),(-8,-8);
Query OK, 8 rows affected
Time: 0.008s
mysql root@127.0.0.1:test> select * from th partition (`p0`);
+----+----+
| a  | b  |
+----+----+
| -5 | -5 |
+----+----+
1 row in set
Time: 0.006s
mysql root@127.0.0.1:test> select * from th partition (`p1`);
+----+----+
| a  | b  |
+----+----+
| -1 | -1 |
| -6 | -6 |
+----+----+
2 rows in set
Time: 0.006s
mysql root@127.0.0.1:test> select * from th partition (`p2`);
+----+----+
| a  | b  |
+----+----+
| -2 | -2 |
| -7 | -7 |
+----+----+
2 rows in set
Time: 0.005s
mysql root@127.0.0.1:test> select * from th partition (`p3`);
+----+----+
| a  | b  |
+----+----+
| -3 | -3 |
| -8 | -8 |
+----+----+
2 rows in set
Time: 0.005s
mysql root@127.0.0.1:test> select * from th partition (`p4`);
+----+----+
| a  | b  |
+----+----+
| -4 | -4 |
+----+----+
1 row in set

```

Reply:
ok

Reply:
Don't forget to add tests

Reply:
done.